File: app/agents/router.py
import re
from dataclasses import dataclass
from datetime import datetime, timedelta


REGIONS = ['北京', '上海', '广州']


@dataclass(slots=True)
class PlannedToolCall:
    tool_name: str
    arguments: dict


class MessageRouter:
    def route(self, message: str) -> list[PlannedToolCall]:
        if self._should_analyze_operational_anomaly(message):
            return [self._parse_operational_anomaly(message)]
        if '分派给' in message:
            return [self._parse_assign(message)]
        if '备注' in message and ('补充' in message or '添加' in message):
            return [self._parse_add_comment(message)]
        if '升级' in message and re.search(r'\bP[123]\b', message):
            return [self._parse_escalate(message)]
        if '日报' in message or '周报' in message:
            return [PlannedToolCall(tool_name='generate_report', arguments={'report_type': 'daily'})]
        if '超' in message and 'SLA' in message:
            return [self._parse_sla_query(message)]
        ticket_no = self._extract_ticket_no(message)
        if ticket_no and ('详情' in message or '操作记录' in message):
            calls = [PlannedToolCall(tool_name='get_ticket_detail', arguments={'ticket_no': ticket_no})]
            if '操作记录' in message or '备注' in message:
                calls.append(PlannedToolCall(tool_name='get_ticket_comments', arguments={'ticket_no': ticket_no}))
            return calls
        if '发布' in message:
            return [PlannedToolCall(tool_name='get_recent_releases', arguments={})]
        if '退款率' in message and '异常' in message:
            return [self._parse_refund_anomaly(message)]
        if '退款率' in message:
            return [self._parse_refund_metric(message)]
        return [PlannedToolCall(tool_name='generate_report', arguments={'report_type': 'daily'})]

    def _parse_refund_anomaly(self, message: str) -> PlannedToolCall:
        start_date, end_date = self._extract_date_range(message)
        return PlannedToolCall(
            tool_name='find_refund_anomalies',
            arguments={
                'start_date': start_date,
                'end_date': end_date,
                'region': self._extract_region(message),
                'top_k': self._extract_top_k(message, default=5),
            },
        )

    def _parse_refund_metric(self, message: str) -> PlannedToolCall:
        start_date, end_date = self._extract_date_range(message)
        return PlannedToolCall(
            tool_name='query_refund_metrics',
            arguments={
                'start_date': start_date,
                'end_date': end_date,
                'region': self._extract_region(message),
                'category': self._extract_category(message),
            },
        )

    def _parse_operational_anomaly(self, message: str) -> PlannedToolCall:
        return PlannedToolCall(
            tool_name='analyze_operational_anomaly',
            arguments={
                'date': self._extract_single_date(message),
                'region': self._extract_region(message),
            },
        )

    def _parse_sla_query(self, message: str) -> PlannedToolCall:
        target_date = self._extract_single_date(message)
        group_by = 'root_cause' if '原因' in message else None
        if '优先级' in message:
            group_by = 'priority'
        if group_by is None and ('类目' in message or '分类' in message):
            group_by = 'category'
        return PlannedToolCall(
            tool_name='list_sla_breached_tickets',
            arguments={'date': target_date, 'region': self._extract_region(message), 'group_by': group_by},
        )

    def _parse_assign(self, message: str) -> PlannedToolCall:
        ticket_no = self._extract_ticket_no(message)
        assignee_match = re.search(r'分派给([\u4e00-\u9fa5A-Za-z0-9_]+)', message)
        assignee_name = assignee_match.group(1) if assignee_match else '待确认'
        return PlannedToolCall(
            tool_name='propose_assign_ticket',
            arguments={
                'ticket_no': ticket_no,
                'assignee_name': assignee_name,
                'reason': f'根据用户指令将 {ticket_no} 分派给 {assignee_name}',
            },
        )

    def _parse_add_comment(self, message: str) -> PlannedToolCall:
        ticket_no = self._extract_ticket_no(message)
        comment_match = re.search(r'备注[:：]\s*(.+)$', message)
        comment_text = comment_match.group(1) if comment_match else message.split('备注', 1)[-1].strip(' ：:')
        return PlannedToolCall(
            tool_name='propose_add_ticket_comment',
            arguments={
                'ticket_no': ticket_no,
                'comment_text': comment_text,
                'reason': f'根据用户指令为 {ticket_no} 增加备注',
            },
        )

    def _parse_escalate(self, message: str) -> PlannedToolCall:
        ticket_no = self._extract_ticket_no(message)
        priority = re.search(r'\b(P[123])\b', message).group(1)
        return PlannedToolCall(
            tool_name='propose_escalate_ticket',
            arguments={
                'ticket_no': ticket_no,
                'new_priority': priority,
                'reason': f'根据用户指令将 {ticket_no} 升级为 {priority}',
            },
        )

    def _extract_ticket_no(self, message: str) -> str:
        match = re.search(r'T\d{6,}', message)
        return match.group(0) if match else ''

    def _extract_region(self, message: str) -> str | None:
        for region in REGIONS:
            if region in message:
                return region
        return None

    def _extract_category(self, message: str) -> str | None:
        for category in ['生鲜', '餐饮', '酒店', '到店综合']:
            if category in message:
                return category
        return None

    def _extract_top_k(self, message: str, default: int = 5) -> int:
        match = re.search(r'(\d+)\s*个', message)
        return int(match.group(1)) if match else default

    def _extract_date_range(self, message: str) -> tuple[str, str]:
        today = datetime.now().date()
        if '最近7天' in message or '最近 7 天' in message:
            return ((today - timedelta(days=6)).isoformat(), today.isoformat())
        if '上周' in message:
            weekday = today.weekday()
            end = today - timedelta(days=weekday + 1)
            start = end - timedelta(days=6)
            return (start.isoformat(), end.isoformat())
        if '昨天' in message:
            yesterday = today - timedelta(days=1)
            return (yesterday.isoformat(), yesterday.isoformat())
        return ((today - timedelta(days=6)).isoformat(), today.isoformat())

    def _extract_single_date(self, message: str) -> str:
        today = datetime.now().date()
        if '昨天' in message:
            return (today - timedelta(days=1)).isoformat()
        if '今天' in message:
            return today.isoformat()
        return today.isoformat()

    def _should_analyze_operational_anomaly(self, message: str) -> bool:
        keywords = ['归因', '关联分析', '关联一下', '异常原因', '发布影响']
        if any(keyword in message for keyword in keywords):
            return True
        return '退款率' in message and 'SLA' in message

File: internal/app/router.go
package app

import (
	"regexp"
	"strconv"
	"strings"
	"time"
)

var regions = []string{"北京", "上海", "广州"}
var categories = []string{"生鲜", "餐饮", "酒店", "到店综合"}

type PlannedToolCall struct {
	ToolName  string         `json:"tool_name"`
	Arguments map[string]any `json:"arguments"`
}

type MessageRouter struct{}

func (r *MessageRouter) Route(message string, memory map[string]any) []PlannedToolCall {
	memoryState, _ := memory["memory_state"].(map[string]any)
	if r.shouldAnalyzeOperationalAnomaly(message) {
		return []PlannedToolCall{r.parseOperationalAnomaly(message, memoryState)}
	}
	if strings.Contains(message, "分派给") {
		return []PlannedToolCall{r.parseAssign(message, memoryState)}
	}
	if strings.Contains(message, "备注") && (strings.Contains(message, "补充") || strings.Contains(message, "添加")) {
		return []PlannedToolCall{r.parseAddComment(message, memoryState)}
	}
	if strings.Contains(message, "升级") && regexp.MustCompile(`\bP[123]\b`).MatchString(message) {
		return []PlannedToolCall{r.parseEscalate(message, memoryState)}
	}
	if strings.Contains(message, "日报") || strings.Contains(message, "周报") {
		return []PlannedToolCall{{ToolName: "generate_report", Arguments: map[string]any{"report_type": "daily"}}}
	}
	if strings.Contains(message, "超") && strings.Contains(strings.ToUpper(message), "SLA") {
		return []PlannedToolCall{r.parseSLAQuery(message, memoryState)}
	}
	ticketNo := r.extractTicketNo(message, memoryState)
	if ticketNo != "" && (strings.Contains(message, "详情") || strings.Contains(message, "操作记录")) {
		calls := []PlannedToolCall{{ToolName: "get_ticket_detail", Arguments: map[string]any{"ticket_no": ticketNo}}}
		if strings.Contains(message, "操作记录") || strings.Contains(message, "备注") {
			calls = append(calls, PlannedToolCall{ToolName: "get_ticket_comments", Arguments: map[string]any{"ticket_no": ticketNo}})
		}
		return calls
	}
	if strings.Contains(message, "发布") {
		return []PlannedToolCall{{ToolName: "get_recent_releases", Arguments: map[string]any{}}}
	}
	if strings.Contains(message, "退款率") && strings.Contains(message, "异常") {
		return []PlannedToolCall{r.parseRefundAnomaly(message, memoryState)}
	}
	if strings.Contains(message, "退款率") {
		return []PlannedToolCall{r.parseRefundMetric(message, memoryState)}
	}
	return []PlannedToolCall{{ToolName: "generate_report", Arguments: map[string]any{"report_type": "daily"}}}
}

func (r *MessageRouter) parseRefundAnomaly(message string, memoryState map[string]any) PlannedToolCall {
	startDate, endDate := r.extractDateRange(message, memoryState)
	return PlannedToolCall{
		ToolName: "find_refund_anomalies",
		Arguments: map[string]any{
			"start_date": startDate,
			"end_date":   endDate,
			"region":     r.extractRegion(message, memoryState),
			"top_k":      r.extractTopK(message, 5),
		},
	}
}

func (r *MessageRouter) parseRefundMetric(message string, memoryState map[string]any) PlannedToolCall {
	startDate, endDate := r.extractDateRange(message, memoryState)
	return PlannedToolCall{
		ToolName: "query_refund_metrics",
		Arguments: map[string]any{
			"start_date": startDate,
			"end_date":   endDate,
			"region":     r.extractRegion(message, memoryState),
			"category":   r.extractCategory(message, memoryState),
		},
	}
}

func (r *MessageRouter) parseOperationalAnomaly(message string, memoryState map[string]any) PlannedToolCall {
	return PlannedToolCall{
		ToolName: "analyze_operational_anomaly",
		Arguments: map[string]any{
			"date":   r.extractSingleDate(message, memoryState),
			"region": r.extractRegion(message, memoryState),
		},
	}
}

func (r *MessageRouter) parseSLAQuery(message string, memoryState map[string]any) PlannedToolCall {
	groupBy := ""
	if strings.Contains(message, "原因") {
		groupBy = "root_cause"
	}
	if strings.Contains(message, "优先级") {
		groupBy = "priority"
	}
	if groupBy == "" && (strings.Contains(message, "类目") || strings.Contains(message, "分类")) {
		groupBy = "category"
	}
	return PlannedToolCall{
		ToolName: "list_sla_breached_tickets",
		Arguments: map[string]any{
			"date":     r.extractSingleDate(message, memoryState),
			"region":   r.extractRegion(message, memoryState),
			"group_by": groupBy,
		},
	}
}

func (r *MessageRouter) parseAssign(message string, memoryState map[string]any) PlannedToolCall {
	ticketNo := r.extractTicketNo(message, memoryState)
	match := regexp.MustCompile(`分派给([\p{Han}A-Za-z0-9_]+)`).FindStringSubmatch(message)
	assigneeName := "待确认"
	if len(match) > 1 {
		assigneeName = match[1]
	}
	return PlannedToolCall{
		ToolName: "propose_assign_ticket",
		Arguments: map[string]any{
			"ticket_no":     ticketNo,
			"assignee_name": assigneeName,
			"reason":        "根据用户指令将 " + ticketNo + " 分派给 " + assigneeName,
		},
	}
}

func (r *MessageRouter) parseAddComment(message string, memoryState map[string]any) PlannedToolCall {
	ticketNo := r.extractTicketNo(message, memoryState)
	commentText := strings.TrimSpace(strings.Trim(strings.SplitN(message, "备注", 2)[1], " ：:"))
	match := regexp.MustCompile(`备注[:：]\s*(.+)$`).FindStringSubmatch(message)
	if len(match) > 1 {
		commentText = strings.TrimSpace(match[1])
	}
	return PlannedToolCall{
		ToolName: "propose_add_ticket_comment",
		Arguments: map[string]any{
			"ticket_no":    ticketNo,
			"comment_text": commentText,
			"reason":       "根据用户指令为 " + ticketNo + " 增加备注",
		},
	}
}

func (r *MessageRouter) parseEscalate(message string, memoryState map[string]any) PlannedToolCall {
	ticketNo := r.extractTicketNo(message, memoryState)
	match := regexp.MustCompile(`\b(P[123])\b`).FindStringSubmatch(message)
	priority := ""
	if len(match) > 1 {
		priority = match[1]
	}
	return PlannedToolCall{
		ToolName: "propose_escalate_ticket",
		Arguments: map[string]any{
			"ticket_no":    ticketNo,
			"new_priority": priority,
			"reason":       "根据用户指令将 " + ticketNo + " 升级为 " + priority,
		},
	}
}

func (r *MessageRouter) extractTicketNo(message string, memoryState map[string]any) string {
	match := regexp.MustCompile(`T\d{6,}`).FindString(message)
	if match != "" {
		return match
	}
	if r.isFollowupReference(message) {
		if value, ok := memoryState["last_ticket_no"].(string); ok {
			return value
		}
	}
	return ""
}

func (r *MessageRouter) extractRegion(message string, memoryState map[string]any) any {
	for _, region := range regions {
		if strings.Contains(message, region) {
			return region
		}
	}
	if r.isFollowupReference(message) {
		if value, ok := memoryState["last_region"].(string); ok && strings.TrimSpace(value) != "" {
			return value
		}
	}
	return nil
}

func (r *MessageRouter) extractCategory(message string, memoryState map[string]any) any {
	for _, category := range categories {
		if strings.Contains(message, category) {
			return category
		}
	}
	if r.isFollowupReference(message) {
		if value, ok := memoryState["last_category"].(string); ok && strings.TrimSpace(value) != "" {
			return value
		}
	}
	return nil
}

func (r *MessageRouter) extractTopK(message string, defaultValue int) int {
	match := regexp.MustCompile(`(\d+)\s*个`).FindStringSubmatch(message)
	if len(match) < 2 {
		return defaultValue
	}
	value, err := strconv.Atoi(match[1])
	if err != nil {
		return defaultValue
	}
	return value
}

func (r *MessageRouter) extractDateRange(message string, memoryState map[string]any) (string, string) {
	today := time.Now().Format("2006-01-02")
	now, _ := time.Parse("2006-01-02", today)
	if strings.Contains(message, "最近7天") || strings.Contains(message, "最近 7 天") {
		return now.AddDate(0, 0, -6).Format("2006-01-02"), now.Format("2006-01-02")
	}
	if strings.Contains(message, "上周") {
		weekday := int(now.Weekday())
		if weekday == 0 {
			weekday = 7
		}
		end := now.AddDate(0, 0, -weekday)
		start := end.AddDate(0, 0, -6)
		return start.Format("2006-01-02"), end.Format("2006-01-02")
	}
	if strings.Contains(message, "昨天") {
		yesterday := now.AddDate(0, 0, -1).Format("2006-01-02")
		return yesterday, yesterday
	}
	if r.isFollowupReference(message) {
		if dateRange, ok := memoryState["last_date_range"].(map[string]any); ok {
			startDate, _ := dateRange["start_date"].(string)
			endDate, _ := dateRange["end_date"].(string)
			if startDate != "" && endDate != "" {
				return startDate, endDate
			}
		}
	}
	return now.AddDate(0, 0, -6).Format("2006-01-02"), now.Format("2006-01-02")
}

func (r *MessageRouter) extractSingleDate(message string, memoryState map[string]any) string {
	today := time.Now().Format("2006-01-02")
	now, _ := time.Parse("2006-01-02", today)
	if strings.Contains(message, "昨天") {
		return now.AddDate(0, 0, -1).Format("2006-01-02")
	}
	if strings.Contains(message, "今天") {
		return now.Format("2006-01-02")
	}
	if r.isFollowupReference(message) {
		if value, ok := memoryState["last_date"].(string); ok && value != "" {
			return value
		}
	}
	return now.Format("2006-01-02")
}

func (r *MessageRouter) shouldAnalyzeOperationalAnomaly(message string) bool {
	keywords := []string{"归因", "关联分析", "关联一下", "异常原因", "发布影响"}
	for _, keyword := range keywords {
		if strings.Contains(message, keyword) {
			return true
		}
	}
	return strings.Contains(message, "退款率") && strings.Contains(strings.ToUpper(message), "SLA")
}

func (r *MessageRouter) isFollowupReference(message string) bool {
	keywords := []string{"他", "它", "这个工单", "该工单", "刚才", "刚刚", "上一个", "那个", "继续"}
	for _, keyword := range keywords {
		if strings.Contains(message, keyword) {
			return true
		}
	}
	return false
}
